compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs

```diff
@@ -378,101 +380,137 @@ typeOfTmLam (var, typ) body = do
 typeOfTyLam :: MonadGamma m => (TypeVarName, Kind) -> Expr -> m Type
 typeOfTyLam (tvar, kind) expr = TForall (tvar, kind) <$> introTypeVar tvar kind (typeOf expr)
 
-typeOfAlts :: MonadGamma m => (CaseAlternative -> m Type) -> [CaseAlternative] -> m Type
+-- | Type to track which constructor ranks have be covered in a pattern matching.
+data MatchedRanks = AllRanks | SomeRanks !(Set.Set Int)
+
+emptyMR :: MatchedRanks
+emptyMR = SomeRanks Set.empty
+
+singletonMR :: Int -> MatchedRanks
+singletonMR k = SomeRanks (Set.singleton k)
+
+unionMR :: MatchedRanks -> MatchedRanks -> MatchedRanks
+unionMR AllRanks _ = AllRanks
+unionMR _ AllRanks = AllRanks
+unionMR (SomeRanks ks) (SomeRanks ls) = SomeRanks (ks `Set.union` ls)
+
+missingMR :: MatchedRanks -> Int -> Maybe Int
+missingMR AllRanks _ = Nothing
+missingMR (SomeRanks ks) n
+    | Set.size ks == n = Nothing
+    | otherwise = Set.lookupMin (Set.fromList [0..n-1] `Set.difference` ks)
+
+typeOfAlts :: MonadGamma m => (CaseAlternative -> m (MatchedRanks, Type)) -> [CaseAlternative] -> m (MatchedRanks, Type)
 typeOfAlts f alts = do
-    ts <- traverse f alts
+    (ks, ts) <- unzip <$> traverse f alts
     case ts of
         [] -> throwWithContext EEmptyCase
         t:ts' -> do
             forM_ ts' $ \t' -> unless (alphaType t t') $
                 throwWithContext ETypeMismatch{foundType = t', expectedType = t, expr = Nothing}
-            pure t
+            let k = foldl' unionMR emptyMR ks
+            pure (k, t)
 
-typeOfAltsVariant :: MonadGamma m => Qualified TypeConName -> [Type] -> [TypeVarName] -> [(VariantConName, Type)] -> [CaseAlternative] -> m Type
+typeOfAltsVariant :: MonadGamma m => Qualified TypeConName -> [Type] -> [TypeVarName] -> [(VariantConName, Type)] -> [CaseAlternative] -> m (MatchedRanks, Type)
 typeOfAltsVariant scrutTCon scrutTArgs tparams cons =
     typeOfAlts $ \(CaseAlternative patn rhs) -> case patn of
         CPVariant patnTCon con var
           | scrutTCon == patnTCon -> do
-            conArgType <- match _Just (EUnknownDataCon con) (con `lookup` cons)
+            (conRank, conArgType) <- match _Just (EUnknownDataCon con) (con `lookupWithIndex` cons)
```

Comment:
Yeah, I agree that this is bad and needs fixing. Since the issue is not coming from this change but has been there before, I'm inclined to fix it in a separate PR. I would fix this in a more general fishing and replace
```haskell
data DataCons
  = DataRecord  ![(FieldName, Type)]
  | DataVariant ![(VariantConName, Type)]
  | DataEnum ![VariantConName]
```
by
```haskell
data DataCons
  = DataRecord  !(Vector (FieldName, Type)) !(Map FieldName Int)
  | DataVariant !(Vector (VariantConName, Type)) !(Map VariantConName Int)
  | DataEnum !(Vector VariantConName) !(Map VariantConName Int)
```
At least for records and variants we have lots of linear lookups all over the type checker. Since values of type `DataCons` are read way more often than they are written, this sort of caching seems reasonable to me. Or do you think that's a bad idea?

Reply:
That seems reasonable, though I'd prefer these "association vector, index map" pairs were wrapped up in a generic type (similar to DA.NameMap) that maintains the invariant we want.

Reply:
Yeah, that's what I was planning to do. I'll take care of it. 😄